### PR TITLE
[Merged by Bors] - Limit snappy input stream

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

N/A

## Proposed Changes

This PR limits the length of the stream received by the snappy decoder to be the maximum allowed size for the received rpc message type. Also adds further checks to ensure that the length specified in the rpc [encoding-dependent header](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/p2p-interface.md#encoding-strategies) is within the bounds for the rpc message type being decoded.

### CODE REVIEW

File: beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs [hunk 1]
```diff
@@ -110,17 +111,24 @@ impl<TSpec: EthSpec> Decoder for SSZSnappyInboundCodec<TSpec> {
 
         let length = self.len.expect("length should be Some");
 
-        // Should not attempt to decode rpc chunks with length > max_packet_size
-        if length > self.max_packet_size {
+        // Should not attempt to decode rpc chunks with `length > max_packet_size` or not within bounds of
+        // packet size for ssz container corresponding to `self.protocol`.
+        let ssz_limits = self.protocol.rpc_request_limits();
+        if length > self.max_packet_size || length > ssz_limits.1 || length < ssz_limits.0 {
```
Comment (protolambda) [inline beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs:117]: When `ssz_limits.1` is 0, the length shouldn't have been read in the first place. I.e. if a metadata request (which has no contents) is received, it shouldn't contain any bytes.

**Review by protolambda — COMMENTED** [review]
Nice test and good changes, but think I found a minor bug in the same code. Metadata requests should be 0 bytes, but it looks like the length bytes are still read.

File: beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs (hunk 1 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs:117]: @protolambda metadata requests do not reach the decoder. They return early here
https://github.com/sigp/lighthouse/blob/56ffe91f9093605737ca1c676ea3493a75b2cfe6/beacon_node/eth2_libp2p/src/rpc/protocol.rs#L224-L226

**Review by AgeManning — COMMENTED** [review]
This looks good. Thanks! 

Although a little more effort, I think we might benefit from implementing an RPC Limit struct type with `max` and `min` fields rather than the tuple. 

I think it will make the code a little more readable and hopefully prevent mixing up the tuple indexes when doing things like `length > limit.1 || length < limit.0`

What do you reckon?

**Comment by pawanjay176** [comment]
Agree this is ugly. Will make the changes :)

**Review by AgeManning — APPROVED** [review]

**Pull request closed without merging**